include/internal/quic_fc.h

```diff
@@ -0,0 +1,266 @@
+/*
+ * Copyright 2022 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#ifndef OSSL_QUIC_FC_H
+# define OSSL_QUIC_FC_H
+
+# include <openssl/ssl.h>
+# include "internal/time.h"
+
+/*
+ * TX Flow Controller (TXFC)
+ * =========================
+ *
+ * For discussion, see doc/designs/quic-design/quic-fc.md.
+ */
+typedef struct quic_txfc_st QUIC_TXFC;
+
+struct quic_txfc_st {
+    QUIC_TXFC  *parent; /* stream-level iff non-NULL */
+    uint64_t    swm, cwm;
+    char        has_become_blocked;
+};
+
+/*
+ * Initialises a TX flow controller. conn_txfc should be non-NULL and point to
+ * the connection-level flow controller if the TXFC is for stream-level flow
+ * control, and NULL otherwise.
+ */
+int ossl_quic_txfc_init(QUIC_TXFC *txfc, QUIC_TXFC *conn_txfc);
+
+/*
+ * Gets the parent (i.e., connection-level) TX flow controller. Returns NULL if
+ * called on a connection-level TX flow controller.
+ */
+QUIC_TXFC *ossl_quic_txfc_get_parent(QUIC_TXFC *txfc);
+
+/*
+ * Sets the parent (i.e., connection-level) TX flow controller. Not normally
+ * needed. Can only be used on a stream-level TX flow controller. This function
+ * cannot be used to change a flow controller from a stream to a
+ * connection-level flow controller or vice versa. Returns 1 on success.
+ */
+int ossl_quic_txfc_set_parent(QUIC_TXFC *txfc, QUIC_TXFC *conn_txfc);
```

Comment:
Why would we ever want to do this?

Reply:
It didn't seem right to constrain the implementation choices of whatever code wants to use a QUIC_FC. It's possible we'll never need it. I can leave it in and we can refactor it out later if it ends up not being needed, or I could take it out and someone can trivially readd it later if it becomes necessary.

Reply:
I'd prefer to remove it. I don't like code hanging around "just in case we need it", because inevitably it never gets cleared away.